robot/impl/local_robot.go

```diff
@@ -655,9 +649,12 @@ func (r *localRobot) newResource(
 }
 
 func (r *localRobot) updateWeakDependents(ctx context.Context) {
-	// track that we are current in resources up to the latest update time. This will
-	// be used to determine if this method should be called while completing a config.
-	r.lastWeakDependentsRound.Store(r.manager.resources.LastUpdatedTime())
+	// Only run updateWeakDependents when the resource graph's logical clock is
+	// now "later" than the value was when we last ran updateWeakDependents.
+	if r.lastWeakDependentsRound.Load() >= r.manager.resources.CurrLogicalClockValue() {
```

Comment:
I moved all the `lastWeakDependentsRound` logic into `updateWeakDependents`, so we just early return if the current logical clock value is equal to or less than the last time we ran `updateWeakDependents`. I _think_ everything still passes, and this reads as logically equivalent to me.

`if r.lastWeakDependentsRound.Load() >= r.manager.resources.CurrLogicalClockValue()` is _not_ a perfect logical inversion of the previous conditional `if r.lastWeakDependentsRound.Load() <= node.UpdatedAt()`; I used `>=` instead of `>`. I think the original conditional _should_ have been `if r.lastWeakDependentsRound.Load() < node.UpdatedAt()`: if the last time we ran `updateWeakDependents` and the current node have an equivalent value, there's no reason to think the graph has been relevantly updated. So, the inversion _should_ probably be `>=`. Let me know if you disagree.

Reply:
That sounds right to me!

Reply:
This change caused [`TestSensorsServiceReconfigure`](https://github.com/viamrobotics/rdk/blob/main/robot/impl/robot_reconfigure_test.go#L2255) to fail (a sort of tangential test of the weak dependency system). 

Unfortunately, checking whether we should update weak dependents by comparing directly against the current graph logical clock value results in too many weak dependency updates (4 vs 1 for subtests within `TestSensorsServiceReconfigure`). The old conditional would only update weak dependents when _some_ resource had _some_ dependency with an `updateAt` value greater than or equal to the last time we ran `updateWeakDependents`; that's obviously slightly more conservative as it examines only pieces of the graph at a time.

If you're curious: the test failed because the sensors service, which is weakly dependent on movement sensors, still found movement sensors even after they were removed from the robot. I don't really understand how that error could result from _too many_ weak dep updates 🤷🏻 . I was looking into it but found that `updateWeakDependents` is sometimes [recursive](https://github.com/viamrobotics/rdk/blob/main/robot/impl/local_robot.go#L764) (that `getDependencies` can call `updateWeakDependents` again), so I gave up trying to understand this for now.

TL;DR I'm not gonna modify the check because I'm definitely breaking something. This PR will remain only comment + lock additions 🙏🏻 .

Reply:
I'm in agreement that the old condition was wrong. Thanks for looking into why your changes weren't working as expected.

> I'd love to mess around with trying to remove the "logical clock" logic entirely and answer the questions "when should we update weak dependents?" and "when do we update weak dependents?", but it's probably too much work for right now, and I'm worried I'm going to start breaking things I don't understand.

I'd also like to explore swapping out the logical clock for something simpler in the future. Now that I see it's purpose, it definitely feels like more of a local "boolean didSomethingChange" variable.

And calling this a logical clock is a bit of a misnomer. It counts like one, but it's not really* being used to establish any happens-before relationships. The naming comes off as a bit intimidating and inaccessible.

Reply:
> Now that I see it's purpose, it definitely feels like more of a local "boolean didSomethingChange" variable.

> And calling this a logical clock is a bit of a misnomer. It counts like one, but it's not really* being used to establish any happens-before relationships. The naming comes off as a bit intimidating and inaccessible.

Well said and agree.